js/stdlib/ts/interfaces.ts

```diff
@@ -1,44 +1,46 @@
-import type { num } from './shared_backend'; // =>
+import type { num, MaybeRep } from './shared_backend'; // =>
 import type { BigNumber } from 'ethers'; // =>
 
 // TODO better typing on these
+type Ty = any
+
 export interface TypeDefs {
-  T_Null: any
-  T_Bool: any
-  T_UInt: any
-  T_Bytes: any
-  T_Address: any
-  T_Digest: any
-  T_Token: any
-  T_Object: any
-  T_Data: any
-  T_Array: any
-  T_Tuple: any
-  T_Struct: any
+  T_Null: Ty
+  T_Bool: Ty
+  T_UInt: Ty
+  T_Bytes: Ty
+  T_Address: Ty
+  T_Digest: Ty
+  T_Token: Ty
+  T_Object: (tyMap: {[key: string]: Ty}) => Ty
+  T_Data: (tyMap: {[key: string]: Ty}) => Ty
+  T_Array: (ty: Ty, size: number) => Ty
+  T_Tuple: (tys: Ty[]) => Ty
+  T_Struct: (nameTyPairs: [string, Ty][]) => Ty
 };
 
 export interface Stdlib_Backend_Shared_User {
-  protect: any,
-  assert: any,
-  Array_set: any,
-  eq: any,
-  ge: any,
-  gt: any,
-  le: any,
-  lt: any,
-  bytesEq: any,
-  digestEq: any,
+  protect: (ty: Ty, v: unknown, m?: string) => unknown
+  assert: (b: boolean, message: string) => void
+  Array_set: <A>(arr: A[], idx: number, val: A) => A[]
+  eq: (n1: num, n2: num) => boolean
+  ge: (n1: num, n2: num) => boolean
+  gt: (n1: num, n2: num) => boolean
+  le: (n1: num, n2: num) => boolean
+  lt: (n1: num, n2: num) => boolean
+  bytesEq: (s1: string, s2: string) => boolean
+  digestEq: (d1: string, d2: string) => boolean
 }
 
 export interface Stdlib_Backend_Shared extends Stdlib_Backend_Shared_User {
-  checkedBigNumberify: any,
-  protect: any,
-  Array_zip: any,
-  mapRef: any,
-  simMapRef: any,
-  simMapSet: any,
-  simMapDupe: any,
-  bytesConcat: any,
+  checkedBigNumberify: (at: string, max: BigNumber, n: any) => BigNumber
+  protect: (t: any, v: unknown, ai?: string) => unknown
+  Array_zip: <A, B>(a1: A[], a2: B[]) => [A, B][]
+  mapRef: <A>(m: {[key: string]: A}, f: string) => MaybeRep<A>
+  simMapRef: (sim_r: unknown, mapi: number, f: string) => MaybeRep<unknown>
```

Comment:
Is this `unknown` because it is too hard to use `ISimRes` here?

Reply:
fsvo "too hard" yes